doc/src/explanation/best-practices.md

```diff
@@ -0,0 +1,1329 @@
+# Best Practices
+
+This page outlines some of the best practices for users of SymPy. The best
+practices here will help avoid some common bugs and pitfalls that can occur
+when using SymPy.
+
+This page primarily focuses on best practices that apply generally to all
+parts of SymPy. Best practices that are specific to certain SymPy submodules
+or functions are outlined in the documentation for those specific functions.
+
+## Basic Usage
+
+(best-practices-defining-symbols)=
+### Defining Symbols
+
+- **The best way to define symbols is using the {func}`~.symbols` function.**
+  The `symbols()` function supports creating one or multiple symbols at a
+  time:
+
+  ```py
+  >>> from sympy import symbols
+  >>> x = symbols('x')
+  >>> a, b, c = symbols('a b c')
+  ```
+
+  Additionally, it supports adding assumptions to symbols
+
+  ```py
+  >>> i, j, k = symbols('i j k', integer=True)
+  ```
+
+  and defining {class}`~.Function` objects:
+
+  ```py
+  >>> from sympy import Function
+  >>> f, g, h = symbols('f g h', cls=Function)
+  ```
+
+  It also supports shorthands for defining many numbered symbols at once:
+
+  ```py
+  >>> symbols('x:10')
+  (x0, x1, x2, x3, x4, x5, x6, x7, x8, x9)
+  ```
+
+  There also exist some alternatives to {func}`~.symbols`. Two alternatives
+  are using the {class}`~.Symbol` constructor directly and {mod}`sympy.abc`.
+  These are both fine, but they are less general than {func}`~.symbols`, so
+  just using `symbols()` is generally preferred. {class}`~.Symbol` only
```

Comment:
Is there a reason `symbols()` is preferred? I recall seeing the use of `Symbol()` in user code discouraged over the years, but never knew why. This doesn't really explain why.

Is it that we want to promote using `symbols()` in our documentation just for style sake? After reading this I'm guessing that all three methods should likely be equally promoted.

Reply:
There is another reason why many people will not want to use symbols which is that it cannot be typed properly because it returns different types based on the value of its argument:
```python
In [1]: type(symbols('x'))
Out[1]: sympy.core.symbol.Symbol

In [2]: type(symbols('x y'))
Out[2]: tuple
```
The `cls=` argument could be typed properly but distinguishing strings like "x" vs "x y" is not something that any type system can do.

Also `Symbol` is more efficient if you just want to make one symbol.

This is sort of mentioned below "if you want to be sure you are only creating one symbol" but it might be worth clarifying why you might want to ensure this: for example if the symbol name is a string that is passed through some end user API then the user might pass a string with spaces e.g. `Matrix([[1,2],[3,4]]).charpoly('long name')`

Reply:
Symbol is fine. I just think it's simpler to have one function that can create one symbol or multiple symbols instead of one function that creates one symbol and another function that creates multiple symbols. With `symbols()`, you basically only ever need the one function, which simplifies things.

Reply:
> you basically only ever need the one function, which simplifies things.

I just stopped using that in production code because `symbols` is unable to infer the types.
In the production code, you anyway are encouraged to do explicit imports from SymPy, so combined that you need to explicitly list you use in SymPy, I rather find that argument opinionated that that using one function having everything combined is really easier.

Reply:
On the other hand, I'd like to listen to the opinion about if `symbols` is used inside for-loop sounds awkward or should be best practice.

Reply:
> Also if assumptions are involved, symbols doesn't make things easier because it sets assumptions simultaneously. And you eventually need to write one symbol for each lines to give according assumptions

In that case you end up with something like what is used in isympy. Different lines for each type of symbol. In my experience, most people just set everything to be real or positive anyway.

```py
>>> from sympy import *
>>> x, y, z, t = symbols('x y z t')
>>> k, m, n = symbols('k m n', integer=True)
>>> f, g, h = symbols('f g h', cls=Function)
```

> On the other hand, I'd like to listen to the opinion about if symbols is used inside for-loop sounds awkward or should be best practice.

I'm not following what you are asking here. What do you mean by using symbols in a for loop?

Reply:
> I'm not following what you are asking here. What do you mean by using symbols in a for loop?

I would rather mean in other context, like building a web app where user can give any kind of equation. I don't use symbols because it parses the spacings.

I don't think that it's worth discussing whether to prefer symbols or symbol. 

In fact, the way how Symbol works behind is one of the easiest things that people can understand.
And it's better to let them trust their own sketches of how it works, than recommending something.

(I have other point of view that programming with type constructor as parameters like cls=Function is not becoming better nowadays because it limits type inference, but that's other topic)

Reply:
Before `symbols`, the common practice was to use `map(Symbol, 'xyz')`. But this is harder to read, and breaks down if you want symbol names with more than one character.

> In fact, the way how Symbol works behind is one of the easiest things that people can understand.

That's really not my experience, especially for people who are new to SymPy (and Python). The way symbols work in SymPy is surprising, especially for people coming from other systems where symbols are defined automatically. Even things like the fact that the variable name has no bearing on the symbol name throws a lot of people (people really have to work with Python for some time before they really *get* how variables work). 

I've seen people use things like `x = S('x')` to create a symbol, which is definitively **not** what you should be doing. Or people just parse full strings of expressions and try to do string expression manipulation (there are other sections in this guide about that). 

For these new users, `symbols` is the best choice.

I agree that if you have programmatic input, you should probably not be using `symbols`. That doesn't apply to 99% of SymPy users though.

Anyway, I'll probably rewrite this section a bit to be a little clearer on these points.

Reply:
I've rewritten this section.